lib/elasticsearch/elastic-query-builder.js

```diff
@@ -0,0 +1,609 @@
+  **/
+  boostOnItemShelfmarkPrefix (boost = 10) {
+    this.query.addShould(prefixMatch('items.shelfMark.raw', boost))
+  }
+
+  /**
+  * Boost bibs with any strong identifier match
+  */
+  boostOnIdentifiers () {
+    const q = this.request.querySansQuotes()
+
+    this.query.addShoulds([
+      prefixMatch('identifierV2.value', q, 10),
+      termMatch('uri', q, 10),
+      prefixMatch('idIsbn', q, 10),
+      prefixMatch('idLccn', q, 10),
+      prefixMatch('idOclc', q, 10),
+      prefixMatch('shelfMark.raw', q, 10),
+      // I would like this to be a prefix, but doesn't seem to work:
+      termMatch('items.identifierV2.value', q),
+      prefixMatch('items.shelfMark.raw', q)
+    ])
+
+    if (/\d{6,}/.test(this.request.querySansQuotes())) {
+      this.query.addShould(prefixMatch('idIsbn_clean', q))
+    }
+  }
+
+  /**
+  * Require bibs have strong tokenized match against given list of fields.
+  */
+  requireMultiMatch (fields, value = this.request.params.q) {
+    const should = []
+
+    // We have an array of fields to match.
+    // Seperate the root-level fields from nested fields by building an object like this:
+    //   {
+    //     _root: [ 'fieldName1', 'fieldName2' ],
+    //     nestedName1: { 'nestedName1.nestedProperty1', 'nestedName1.nestedProperty2' }
+    //   }
+    const fieldMap = fields.reduce((map, fieldName) => {
+      // Handle query conditional field
+      if (typeof fieldName === 'object' && fieldName.on) {
+        if (!fieldName.on(value)) return map
+        fieldName = fieldName.field
+      }
+
+      // Most fields will be matched at root level:
+      let nestedName = '_root'
+      // Any field starting with the following is a nested field:
+      if (['items'].indexOf(fieldName.split('.').shift()) >= 0) {
+        nestedName = fieldName.split('.').shift()
+      }
+      if (!map[nestedName]) map[nestedName] = []
+      map[nestedName].push(fieldName)
+      return map
+    }, { _root: [] })
+
+    should.push({
+      multi_match: namedQuery({
+        fields: fieldMap._root,
+        query: escapeQuery(value),
+        type: 'most_fields',
+        // For 4+ term queries, require 75% of terms:
+        minimum_should_match: '3<75%'
+      }, 'bib multi_match')
+    })
+
+    // Add nested queries (if any) to things that *should* match:
+    Object.keys(fieldMap)
+      .filter((nestedName) => nestedName !== '_root')
+      .forEach((nestedName) => {
+        should.push({
+          nested: {
+            path: nestedName,
+            query: {
+              multi_match: {
+                fields: fieldMap[nestedName],
+                query: escapeQuery(value),
+                type: 'phrase'
+              }
+            }
+          }
+        })
+      })
+
+    this.query.addMust(should.length === 1 ? should[0] : { bool: { should } })
+  }
+
+  /**
+  * Examine request for user-filters. When found, add them to query.
+  */
+  applyFilters () {
```

Comment:
Can this method be broken down any more?

Reply:
I agree it should be, but this specific code is unchanged from previous form (just moved). So I'd like to keep that as a project for a future refactor if that's alright.